---
title: "GWAS"
author: "Trevor Manz"
date: "5/19/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggrepel)
library(patchwork)

set.seed(42)

#### Data utils

read_sheet <- function(name) {
  fs::path("data", "supplementary_tables.xlsx") %>% readxl::read_excel(name)
}

# Load results from Zhou et al, NatGen 201
zhou_results <- function() {
  data <- read_sheet("S5_LDL_TWAS") %>%
    # Drop unused cols
    select(!genetype) %>%
    # Make table tidy
    mutate(chr = as.character(chr)) %>%
    mutate(across(effect_size_PrediXcan:PFDR_JTI, as.numeric)) %>%
    pivot_longer(
      !c(genename, geneid, chr, left, right),
      names_to = c(".value", "method"),
      names_pattern = "(.*)_(PrediXcan|JTI)"
    ) %>%
    # arrange by method
    arrange(desc(method)) %>%
    rename(pfdr = PFDR) %>%
    mutate(bp = left) %>%
    drop_na()
}

# Load local GWAS result
read_result <- function(ukbb_id, method, tissue) {
  fs::path("results", paste0(ukbb_id, "-", method, "_", tissue, ".csv")) %>%
    read_csv() %>%
    # Match columns to zhou results
    rename(geneid = gene, genename = gene_name) %>%
    # Add new cols
    mutate(method = method, pfdr = p.adjust(pvalue, method = "fdr"))
}

# Merge results from multiple methods
read_results <- function(ukbb_id, tissue, methods = c("PrediXcan", "JTI", "UTMOST")) {
  methods %>%
    map(function(m) read_result(ukbb_id, m, tissue)) %>%
    bind_rows()
}


#### Plotting utils

COLORS <- c("#B4D88B", "#A6CEE2", "#34A048", "#1F78B4")
FDR <- 0.05

manhattan_base <- function(res) {
  # Adapted from https://www.r-graph-gallery.com/101_Manhattan_plot.html
  # get cumulative position
  res <- res %>%
    group_by(chr) %>%
    summarise(chr_len = max(bp)) %>%
    arrange(as.numeric(chr)) %>%
    mutate(tot = cumsum(chr_len) - chr_len) %>%
    select(-chr_len) %>%
    left_join(res, ., by = "chr") %>%
    mutate(bp_cum = tot + bp)

  axisdf <- res %>%
    group_by(chr) %>%
    summarize(center = (max(bp_cum) + min(bp_cum)) / 2) %>%
    arrange(as.numeric(chr))

  ggplot(res, aes(x = bp_cum, y = -log10(pvalue))) +
    scale_x_continuous(
      label = c(1:15, "", 17, "", 19, "", 21, ""),
      breaks = axisdf$center
    ) +
    scale_y_log10(
      breaks = c(1, 2, 3, 5, 10, 20, 30, 50, 100, 200, 300, 500),
      limits = c(0.7, 600),
      expand = c(0.04, 0)
    ) +
    theme_bw() +
    theme(
      panel.border = element_blank(),
      panel.grid = element_blank(),
      legend.position = "top"
    ) +
    labs(x = "Chromosome", y = "-log10[P]", color = "")
}

sig_genes_bars <- function(results, fdr = FDR, colors = COLORS) {
  results %>%
    filter(pfdr < fdr) %>%
    ggplot(aes(fct_rev(method), fill = paste(known, method))) +
    geom_bar(stat = "count") +
    geom_text(stat = "count", aes(label = ..count..), vjust = -1) +
    scale_y_continuous(limits = c(0, 700)) +
    facet_wrap(~ !known) +
    ylab("Number of significant genes (p FDR < 0.05)") +
    theme_bw() +
    theme(
      strip.text.x = element_blank(),
      panel.border = element_blank(),
      panel.grid = element_blank(),
      axis.title.x = element_blank(),
      axis.text.x = element_text(angle = 45, hjust = 1)
    ) +
    guides(fill = FALSE) +
    scale_fill_manual(values = colors)
}

manhatten_labeled_fdr <- function(results, fdr = FDR, colors = COLORS, gray_below = FALSE) {
  if (gray_below) {
    below_pts <- geom_point(data = ~ filter(.x, pfdr > fdr), color = "lightgray")
  } else {
    below_pts <- geom_point(data = ~ filter(.x, pfdr > fdr), aes(color = paste("Addtional", method)))
  }
  manhattan_base(results) +
    geom_point(data = ~ filter(.x, pfdr < fdr & !known), aes(color = paste("Addtional", method))) +
    geom_point(data = ~ filter(.x, pfdr < fdr & known), aes(color = paste("Known", method))) +
    below_pts +
    geom_label_repel(
      data = ~ filter(.x, pfdr < fdr & known),
      aes(label = genename, color = paste("Known", method)),
      size = 3,
      alpha = 0.9,
      show.legend = FALSE,
      box.padding = 0.5
    ) +
    scale_color_manual(values = colors)
}

# Creates barplot from paper
fig4 <- function(results, title = NULL, colors = COLORS, gray_below = FALSE) {
  if (!is.null(title)) title <- ggtitle(title)
  m <- manhatten_labeled_fdr(results, colors = colors, gray_below = gray_below) + title
  bars <- sig_genes_bars(results, colors = colors)
  m + bars + plot_layout(widths = c(5, 1))
}
```

# Manhattan plot for LDL-C level from the application of PrediXcan and JTI models in liver to UK Biobank LDL-C GWAS summary statistics.
```{r, warning=FALSE, message=FALSE}
# Load known reference genes
known_genes <- read_sheet("S4_LDL_known_genes") %>% pull(genename)

# Load results from Zhou et. al Nat Gen 2020
zhou <- zhou_results() %>% mutate(known = genename %in% known_genes)

# Create lookup table for genes
genes <- zhou %>% distinct(geneid, genename, known, chr, bp)

# Load association results for LDL & Liver
ldl <- read_results(ukbb_id = "30780_irnt", tissue = "Liver") %>% inner_join(genes)
```

```{r, warning=FALSE, fig.fullwidth = TRUE, fig.width=10, fig.height=10}
fig4(ldl %>% filter(method != "UTMOST"), title = "Manz 2021") / fig4(zhou, title = "Zhou et. al, Nature Gen 2020")
```
  
# Annotate known genes that don't match Zhou et al.
```{r, warning=FALSE, fig.fullwidth = TRUE, fig.width=10, fig.height=10}
filter_missing <- function(x) filter(x, method == "JTI" & genename %in% c("LIPA", "APOC2"))

manz2 <- (manhatten_labeled_fdr(ldl %>% filter(method != "UTMOST"), gray_below = TRUE) +
  geom_label_repel(data = filter_missing, aes(label = genename), size = 3, show.legend = FALSE, box.padding = 0.5) +
  geom_point(data = filter_missing) +
  ggtitle("Manz 2021")
) + sig_genes_bars(ldl %>% filter(method != "UTMOST")) + plot_layout(widths = c(5, 1))

zhou2 <- (manhatten_labeled_fdr(zhou, gray_below = TRUE) +
  geom_point(data = filter_missing) +
  ggtitle("Zhou et. al, Nature Gen 2020")
) + sig_genes_bars(zhou) +
  plot_layout(widths = c(5, 1))

manz2 / zhou2
```

# Compare with JTI with UTMOST

```{r, warning=FALSE, fig.fullwidth = TRUE, fig.width=10, fig.height=6}
fig4(ldl %>% filter(method != "PrediXcan"), colors = c("#B4D88B", "#FCAE6B", "#34A048", "#E6550E"))
```

```{r, warning=FALSE, fig.fullwidth = TRUE, fig.width=10, fig.height=6}
fig4(
  results = ldl,
  colors = c("#B4D88B", "#A6CEE2", "#FCAE6B", "#34A048", "#1F78B4", "#E6550E"),
  gray_below = TRUE
)
```
