# ðŸ““ 

This repository contains an entire end-to-end workflow to reproduce gene-level
association studies described in:

> Zhou, D., Jiang, Y., Zhong, X. et al. A unified framework for 
> joint-tissue transcriptome-wide association and Mendelian 
> randomization analysis. _Nat Genet_ 52 (2020). https://doi.org/10.1038/s41588-020-0706-2

The workflow is described in and executed by [Snakemake](https://snakemake.readthedocs.io/en/stable/index.html), 
a management system for scalable reproducible analysis pipelines.

## Requirements

- Python 3.9
- `conda` - https://docs.conda.io/en/latest/
- `R>=4.0` including `tidyverse` and `bookdown` packages (the latter for creating `summary.pdf`)

## Running workflow

Clone this repository to reproduce associations and generate `summary.pdf`. Associations results
are generated by [`SPrediXcan.py`](https://github.com/hakyimlab/MetaXcan) and saved under
`results/{ukbb_id}-{model}_{tissue}.csv`.

```bash
$ git clone https://github.com/manzt/zhou-et-al-natgen-2020 && cd zhou-et-al-natgen-2020
$ conda env create --file environment.yml
$ snakemake --cores all # downloads all data, runs associations, and generates `summary.pdf`
```

Running the complete workflow will download and organize **all** the input data
necessary to reproduce the gene-level association results described in Table S7,

- GWAS summary statistics from UKBB
- Pretrained Transcriptome Prediction Model databases (JTI, PrediXcan, UTMOST)
- Covariance matrices of SNps within each gene model

The data are organized in the following directory structure, and associations are 
generated using `SPrediXcan.py` from [`MetaXcan`](https://github.com/hakyimlab/MetaXcan).

```bash
data/
â”œâ”€â”€ covariances/
â”œâ”€â”€ GWAS/
â”œâ”€â”€ weights/
â””â”€â”€ supplementary_tables.xlsx
```

## Individual associations

Running all the associations can take some time. If you are interested in running 
individual associations, you may run `SPrediXcan.py` for individual GWAS summary 
statistics / model-tissue -specific weights explicitly by matching the following
pattern with a snakmake wildcard `results/{ukbb_id}-{model}_{tissue}.csv`.

For example,

```bash
$ snakemake --cores all results/30740_irnt-UTMOST_Muscle_Skeletal.csv
```

will only perform the steps necessary to produce the gene-level association output 
for Glucose (`37040_irnt`) using the pretrained `UTMOST` weights for Muscle Skeletal tissue.
